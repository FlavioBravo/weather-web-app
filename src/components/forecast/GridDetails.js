import React from 'react';
import PropTypes from 'prop-types';

export const GridDetails = ({ hourlyItem }) => {
    return (
        <div>
            <br />
            <hr />
            <br />
            <h3>Hourly Forecast</h3>
            <br />
            <div className="gridDetails_container">
            {
            hourlyItem.map( hour  => (
                <p className="gridDetails_linespace" key={hour.dt} >Temperature: {hour.temp} ° | Human Perception: {hour.feels_like} ° | Humidity: {hour.humidity} % | Clouds: {hour.clouds} % | Pressure: {hour.pressure} hPa | Uvi: {hour.uvi} UV</p>
            ))
            }
            </div>
        </div>
    )
}

GridDetails.propTypes = {
    hourlyItem: PropTypes.array.isRequired
}