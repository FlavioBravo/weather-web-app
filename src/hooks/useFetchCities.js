import { useState, useEffect } from 'react'
import { getCities } from '../helpers/getCities';


export const useFetchCities = ( city ) => {
    
    const [state, setState] = useState({
        data: [],
        loading: true
    });

    useEffect( () => {

        getCities( city )
            .then( imgs => {
                
                setState({
                    data: imgs,
                    loading: false
                });
            })

    }, [city])




    return state;


}