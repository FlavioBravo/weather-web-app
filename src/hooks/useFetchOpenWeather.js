import { useState, useEffect } from 'react';
import { get5daysForecast } from '../helpers/get5daysForecast';


export const useFetchOpenWeather = ( cityId ) => {
    
    const [state, setState] = useState({
        data: [],
        loading: true
    });

    useEffect( () => {
        let isSubscribed = true;
        get5daysForecast( cityId )
            .then( forecast => {
                if (isSubscribed) {
                    setState({
                        data: forecast,
                        loading: false
                    });
                }
            })
        return () => isSubscribed = false;
    }, [cityId])

    return state;

}