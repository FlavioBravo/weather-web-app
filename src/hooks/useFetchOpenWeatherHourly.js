import { useState, useEffect } from 'react';
import { getHourlyForecast } from '../helpers/getHourlyForecast';


export const useFetchOpenWeatherHourly = ( lat, lon ) => {
    
    const [state, setState] = useState({
        data: [],
        loading: true
    });

    useEffect( () => {
        let isSubscribed = true;
        getHourlyForecast( lat, lon )
            .then( hourly => {
                if (isSubscribed) {
                    setState({
                        data: hourly,
                        loading: false
                    });
                }
            })
        return () => isSubscribed = false;
    }, [lat, lon])

    return state;

}