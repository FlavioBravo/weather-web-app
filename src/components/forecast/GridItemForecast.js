import React from 'react';
import PropTypes from 'prop-types';
import { getDayNameOfWeek } from '../../helpers/utils';
import { useFetchOpenWeatherHourly } from '../../hooks/useFetchOpenWeatherHourly';

export const GridItemForecast = ({ dt, temp_min, temp_max, weather, icon, dt_txt, lat, lon, setHourForecast }) => {

    const dayName = getDayNameOfWeek(dt_txt);
    const dateFormat = dt_txt.split(' ')[0];

    const { data: hourly, loading } = useFetchOpenWeatherHourly(lat, lon);

    const handleForecast = () => {
        setHourForecast(hourly);
    };

    return (
        <div className="animate__animated animate__fadeIn forecast_grid_item" onClick={ handleForecast } >
            <p> { dayName } </p>
            <p> { weather } </p>
            <img src={ icon } alt={ weather } height="80" width="80" />
            <div className="forecast_grid_item_footer">
                <span>{ temp_min }°</span>
                <span>{ temp_max }°</span>
            </div>
            <p> { dateFormat } </p>
        </div>
    )
}

GridItemForecast.propTypes = {
    temp_min: PropTypes.number.isRequired,
    temp_max: PropTypes.number.isRequired,
    weather: PropTypes.string.isRequired,
    icon: PropTypes.string.isRequired,
    dt_txt: PropTypes.string.isRequired,
}