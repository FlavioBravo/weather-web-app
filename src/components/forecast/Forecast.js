import { React, useState } from 'react';
import { useParams } from "react-router-dom";

import { useFetchOpenWeather } from '../../hooks/useFetchOpenWeather';
import { GridDetails } from './GridDetails';
import { GridItemForecast } from './GridItemForecast';

export const Forecast = () => {

    const { cityId } = useParams();

    const { data: daysForecast, loading } = useFetchOpenWeather(cityId);

    const [state, setState] = useState({
        hourlyItem: [],
        show: false
    });
    
    const setHourForecast = (hourly) => {
        if (hourly) {
            setState({
                hourlyItem: [ ...hourly ],
                show: true
            });
        } else {
            setState({
                hourlyItem: [],
                show: false
            });
        }
    };

    return (
        <>
            <h2>Forecast</h2>
            <br />
            <hr />
            <br />
            { loading && <p className="animate__animated animate__flash">Loading</p> }
            <div className="forecast_container">
              {
                  daysForecast.map( day  => (
                    <GridItemForecast
                        key={day.dt_txt}
                        {...day}
                        setHourForecast = { setHourForecast }
                    />
                  ))
              }
          </div>
            {
               state.show && 
               <GridDetails
                    hourlyItem = { state.hourlyItem }
                />
            }

        </>
    )
}
