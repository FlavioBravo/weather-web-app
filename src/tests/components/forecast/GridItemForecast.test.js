import React from 'react';
import '@testing-library/jest-dom';

import { shallow } from 'enzyme';
import { GridItemForecast } from '../../../components/forecast/GridItemForecast';
import { useFetchOpenWeatherHourly } from '../../../hooks/useFetchOpenWeatherHourly';
jest.mock('../../../hooks/useFetchOpenWeatherHourly');


describe('Pruebas en el <GridItemForecast />', () => {

    const day = {
        dt: 1622808000,
        dt_txt: "2021-06-04 12:00:00",
        icon: "http://openweathermap.org/img/wn/04d.png",
        lat: -3.5667,
        lon: -80.4414,
        temp_max: 21.95,
        temp_min: 21.95,
        weather: "Clouds"
    };
    const hourly = [
        {
            dt: 1622224800,
            temp: 299.07,
            feels_like: 299.65,
            pressure: 1011,
            humidity: 74,
            dew_point: 294.08,
            uvi: 10.54,
            clouds: 42,
            visibility: 10000,
            wind_speed: 4.74,
            wind_deg: 302,
            wind_gust: 3.25,
            weather: [
                {
                    id: 802,
                    main: "Clouds",
                    description: "scattered clouds",
                    icon: "03d"
                }
            ],
            pop: 0
        },
        {
            dt: 1622228400,
            temp: 298.91,
            feels_like: 299.5,
            pressure: 1011,
            humidity: 75,
            dew_point: 294.14,
            uvi: 8.82,
            cloud: 45,
            visibility: 10000,
            wind_speed: 5.29,
            wind_deg: 286,
            wind_gust: 4.27,
            weather: [
                {
                    id: 802,
                    main: "Clouds",
                    description: "scattered clouds",
                    icon: "03d"
                }
            ],
            pop: 0.02
        }
    ];
    const setHourForecast = jest.fn();
    

    test('debe de mostrarse correctamente', () => {

        useFetchOpenWeatherHourly.mockReturnValue({
            data: [],
            loading: false
        });

        const wrapper = shallow( <GridItemForecast {...day} setHourForecast={ setHourForecast } /> );
        expect( wrapper ).toMatchSnapshot();
        
    });

    test('debe de tener la imagen igual al url y alt de los props', () => {

        useFetchOpenWeatherHourly.mockReturnValue({
            data: hourly,
            loading: true
        });
        const wrapper = shallow( <GridItemForecast {...day} setHourForecast={ setHourForecast } /> );
        const img = wrapper.find('img');
        
        expect( img.prop('src') ).toBe( day.icon );
        expect( img.prop('alt') ).toBe( day.weather );

        const p = wrapper.find('p');
        expect( p.length ).toBe( 3 );

        const span = wrapper.find('span');
        expect( span.length ).toBe( 2 );

    });
    
    

})