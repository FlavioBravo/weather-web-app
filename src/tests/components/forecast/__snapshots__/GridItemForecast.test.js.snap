// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el <GridItemForecast /> debe de mostrarse correctamente 1`] = `
<div
  className="animate__animated animate__fadeIn forecast_grid_item"
  onClick={[Function]}
>
  <p>
     
    Friday
     
  </p>
  <p>
     
    Clouds
     
  </p>
  <img
    alt="Clouds"
    height="80"
    src="http://openweathermap.org/img/wn/04d.png"
    width="80"
  />
  <div
    className="forecast_grid_item_footer"
  >
    <span>
      21.95
      °
    </span>
    <span>
      21.95
      °
    </span>
  </div>
  <p>
     
    2021-06-04
     
  </p>
</div>
`;
